src/state_machine.zig

```diff
@@ -891,6 +943,211 @@ pub fn StateMachineType(
             self.prefetch_finish();
         }
 
+        fn prefetch_expire_pending_transfers(self: *StateMachine) void {
+            assert(self.scan_result_count == 0);
+            assert(self.prefetch_timestamp >= TimestampRange.timestamp_min);
+            assert(self.prefetch_timestamp != TimestampRange.timestamp_max);
+
+            var scan_buffer = std.mem.bytesAsSlice(
+                Transfer,
+                self.scan_buffer[0..@sizeOf(Transfer)],
+            );
+            // We only need the first result.
+            assert(scan_buffer.len == 1);
+
+            const transfers_groove: *TransfersGroove = &self.forest.grooves.transfers;
+            const scan_builder: *TransfersGroove.ScanBuilder = &transfers_groove.scan_builder;
+
+            // WHERE `expires_at > prefetch_timestamp`
+            // Scanning `Transfers` that _will_ expire, so we can determine the _next_ timestamp
+            // it will require to execute `expire_pending_transfers` again.
+            // We must do it before scanning for expired transfers to reuse the same buffer.
```

Comment:
I am still early in the review, but to help me orient can you explain why we need the `prefetch_expire_pending_transfers` step instead of just jumping right into `prefetch_expire_pending_transfers_scan`?

Naively, I would imagine that we could do something like:
- Fetch the next 8190 transfers to expire.
- If we encounter one of those that shouldn't expire yet, save that timestamp so that we don't pulse again until then.
- If all 8190 transfers need to expire, then we will pulse again.

Reply:
The prefetch work is split into:

1. Scans for a single transfer that will expire next but has not expired yet (e.g., `WHERE expires_at > prefetch_timestamp ORDER BY ASC LIMIT 1`). This will be used to determine when a `pulse` is needed again.

    We do this step first to reuse the same buffer in the next step.

2. Scans for all transfers that have expired (e.g. `WHERE expires_at <= prefetch_timestamp ORDER BY ASC`).

3. Lookup the Debit and Credit accounts for each transfer.

4. Lookup the Posted for each transfer if `constants.verify`, to assert if the transfer wasn't already posted.

> Naively, I would imagine that we could do something like

I haven't considered unbounded fetching the next 8190 transfer to expire ( `WHERE expires_at > 0 ORDER BY ASC`) because we would do unnecessary work, since we need to always "materialize" all 8190 transfers and then when iterating maybe just a handful of them would need to expire.

But it's a nice idea if the Scan accepts a function pointer so the caller can tell when to stop (the first which is `expires_at > prefetch_timestamp`).
So we could do in a single scan the equivalent of what we are currently doing in two separated ones:

```SQL
WHERE expires_at <= prefetch_timestamp ORDER BY ASC
UNION
WHERE expires_at > prefetch_timestamp ORDER BY ASC LIMIT 1
```

Reply:
> But it's a nice idea if the Scan accepts a function pointer so the caller can tell when to stop

Yeah, that sounds like it would be perfect! The two queries are adjacent, so the second one can (usually) benefit from the first's cached blocks, but having a single query would be simpler too.